### DISCUSSION TITLE
Slow latency on first query?

### DISCUSSION THREAD

**Discussion by JonathanHope** [ncruces/go-sqlite3#32] [Q&A] [answered]
Love the idea behind this!

I've noticed some surprising latency the first time a query is run (`db.Query`). On my machine it takes about 400ms. Every call to `db.Query` afterwards returns in comparable time to other drivers.

It's like there is some kind of expensive one time setup that occurs at that point. I don't know if this is expected behavior or not.

Thanks!

**Comment by ncruces** [comment]
This is a good question that I'd like to keep highlighted here, rather than in a closed issue.
Hope you don't mind!

**Comment by ncruces** [comment, marked as answer]
Edit: this issue should be resolved since we moved away from `wazero`.

---

Yes, this is expected behaviour, though maybe it should be documented better.

We embed a Wasm version of SQLite. This needs to be compiled to machine code on the first instantiation:
https://github.com/ncruces/go-sqlite3/blob/58c50093dab002d016dad1b4fdf5f3829bda0e20/sqlite.go#L70-L71

[wazero](https://wazero.io) (our WASM runtime) is quite fast at compiling, but the WASM file is also quite big ([1.4Mb](https://github.com/ncruces/go-sqlite3/blob/main/embed/sqlite3.wasm)), hence the slowdown.

There are 3 ways to mitigate this cost:

1. compile at a more appropriate time
2. cache the compilation result
3. slim down SQLite to reduce compilation time

### Compile at a more appropriate time

This should do the job:

```go
if err := sqlite3.Initialize(); err != nil {
	log.Fatal(err)
}
```

You can also do this in a background goroutine.

### Cache the compilation result

You can provide your own [`wazero.RuntimeConfig`](https://pkg.go.dev/github.com/tetratelabs/wazero#RuntimeConfig), by setting [`sqlite3.RuntimeConfig`](https://pkg.go.dev/github.com/ncruces/go-sqlite3#pkg-variables). This allows you to configure the compilation cache, choose between the compiler/interpreter, limit memory, etc…) The cache can speed up first instantiation by an order of magnitude.

### Slim down SQLite to reduce compilation time

If you can do away with some of the [optional features](https://github.com/ncruces/go-sqlite3/blob/main/embed/README.md), one way to improve the situation is to generate your own, slimmed down, [`sqlite3.wasm`](https://github.com/ncruces/go-sqlite3/blob/main/embed/sqlite3.wasm). You can also trade speed for size (`-O3` vs `-Os`).

All you need is [`wasi-sdk`](https://github.com/WebAssembly/wasi-sdk), and [`binaryen`](https://github.com/WebAssembly/binaryen). You can edit [`sqlite_opt.h`](https://github.com/ncruces/go-sqlite3/blob/main/sqlite3/sqlite_opt.h) to remove some of the [extensions](https://github.com/ncruces/go-sqlite3/blob/0ec08c2e740a010c5a0e2e9ee60d1f3eb229fa54/sqlite3/sqlite_cfg.h#L69-L76) (FTS, R*Tree, GeoPoly are good candidates if you don't need them). And tweak compile options in the [build.sh](https://github.com/ncruces/go-sqlite3/blob/main/embed/build.sh) script. To load your own build of SQLite, you can change either the `Binary` or `Path` [variables](https://pkg.go.dev/github.com/ncruces/go-sqlite3#pkg-variables), instead of importing `embed`.

**Comment by infogulch** [comment]
Would it be possible to precompile the WASM binary at go build time, then embed *that* instead of the WASM code? This could eliminate the startup compilation delay. The architecture is known at go build time, so this should be possible in theory (architecture specialization based on the current machine notwithstanding, idk if the wazero compiler does that). I don't see a nice API for this in wazero, though that hasn't stopped us in the past...

Edit: Oh this was discussed in https://github.com/tetratelabs/wazero/issues/1733

Seems not worth it in my case for these reasons mentioned in the thread:

1. the generated machine code is 10x the size of the wasm code (wow! did not realize .wasm was that efficient)
2. wazero *does* specialize machine code gen based on the currently running cpu, and it does not currently allow limiting specialization. this tanks portability
3. no cross-compilation story at all (see 2)
